python-checks/src/main/java/org/sonar/python/checks/cdk/UnrestrictedAdministrationCheck.java

```diff
@@ -43,146 +44,14 @@
-        }
-      });
-  }
-
-  private static boolean isDictionaryWithAttributeBadProtocolEmptyIpAddressAdminPort(DictionaryAsMap map) {
-    return map.hasKeyValuePair(IPPROTOCOL, isString(SENSITIVE_PROTOCOL))
-      && (map.hasKeyValuePair(CIDRIP, isString(EMPTY_IPV4)) || map.hasKeyValuePair(CIDRIPV6, isString(EMPTY_IPV6)))
-      && map.hasSensitivePortRange("fromPort", "toPort");
-  }
-
-  private static boolean isDictionaryWithAttributeInvalidProtocolEmptyIpAddress(DictionaryAsMap map) {
-    return map.hasKeyValuePair(IPPROTOCOL, isString(ANY_PROTOCOL))
-      && (map.hasKeyValuePair(CIDRIP, isString(EMPTY_IPV4)) || map.hasKeyValuePair(CIDRIPV6, isString(EMPTY_IPV6)));
-  }
-
-  // Class to handle Dictionary elements as a Map of ResolvedKeyValuePair, useful in case we need to refer to several elements and not only looking for a specific one.
-  // The keys are all resolved String in CdkUtils.ResolvedKeyValuePair.key
-  static class DictionaryAsMap {
-    Map<String, CdkUtils.ResolvedKeyValuePair> map = new HashMap<>();
-
-    public static DictionaryAsMap build(SubscriptionContext ctx, DictionaryLiteral dictionary) {
-      DictionaryAsMap dict = new DictionaryAsMap();
-
-      List<CdkUtils.ResolvedKeyValuePair> pairs = dictionary.elements().stream()
-        .map(e -> CdkUtils.getKeyValuePair(ctx, e))
-        .filter(Optional::isPresent).map(Optional::get).collect(Collectors.toList());
-
-      for (CdkUtils.ResolvedKeyValuePair pair : pairs) {
-        pair.key.getExpression(isStringLiteral()).map(StringLiteral.class::cast)
-          .ifPresent(key -> dict.map.put(key.trimmedQuotesValue(), pair));
-      }
-      return dict;
-    }
-
-    public boolean hasKeyValuePair(String key, Predicate<Expression> valuePredicate) {
-      return map.containsKey(key) && map.get(key).value.hasExpression(valuePredicate);
-    }
-
-    public Optional<Expression> get(String key, Predicate<Expression> valuePredicate) {
-      if (!map.containsKey(key)) {
-        return Optional.empty();
-      }
-      return map.get(key).value.getExpression(valuePredicate);
-    }
-
-    public void addIssue(String key, String message) {
-      if (map.containsKey(key)) {
-        map.get(key).value.addIssue(message);
-      }
-    }
-
-    public Optional<Long> getArgumentAsLong(String name) {
-      return get(name, isNumericLiteral())
-        .map(NumericLiteral.class::cast)
-        .map(NumericLiteral::valueAsLong);
-    }
-
-    boolean hasSensitivePortRange(String minName, String maxName) {
-      Optional<Long> min = getArgumentAsLong(minName);
-      Optional<Long> max = getArgumentAsLong(maxName);
-
-      if (min.isEmpty() || max.isEmpty()) {
-        return false;
-      }
-
-      return isInInterval(min.get(), max.get(), ADMIN_PORTS);
-    }
+  public void initialize(Context context) {
+    new UnrestrictedAdministrationCheckPartCfnSecurity().initialize(context);
+    new UnrestrictedAdministrationCheckPartConnections().initialize(context);
```

Comment:
I find this very strange: I guess it means we have two different AST visits for the same check? Does it mean we probably need to split this rule?

Reply:
This could be a possibility indeed.

The **CfnSecurity** part is doing checks on :
- aws_cdk.aws_ec2.CfnSecurityGroup
- aws_cdk.aws_ec2.CfnSecurityGroupIngress

While the **Connections** part is doing check on :
- aws_cdk.aws_ec2.Connections
- aws_cdk.aws_ec2.SecurityGroup
- aws_cdk.aws_ec2.*constructs*

The rule is quite big and on our side we splitted the ticket in two part that could be coded independently from each other.
They still do check similar things (ip protocol, ip address, ports...), but on different objects.

So in the end I choose to split the code in two different class for readability, as they don't directly share logic or elements.
WDYT should be done here ?